# Task 4.6 - Combining & Exporting Data - Answers

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

#### 3. In a new notebook, import the orders_products_combined dataframe from the pickle file you just saved

In [2]:
#Create folder path
path=r'C:\Users\Utilizador\OneDrive\Documents\CareerFoundry - Data Analyst\Data Immersion\A4\19-01-2024 Instacart Basket Analysis'

In [3]:
#Import merged orders.csv and orders_products_prior.csv pickle file
df_ords_prods=pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_combined.pkl'))

In [4]:
#check output
df_ords_prods.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             0           0   2539329        1             1   
2             0           0   2539329        1             1   
3             0           0   2539329        1             1   
4             0           0   2539329        1             1   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                   2                  8                     NaN   
1                   2                  8                     NaN   
2                   2                  8                     NaN   
3                   2                  8                     NaN   
4                   2                  8                     NaN   

   is_first_order  product_id  add_to_cart_order  reordered _merge  
0               1         196                  1          0   both  
1               1       14084                  2          0   both  
2               1       12427                  3          0   both  
3               1       26088                  4          0   both  
4               1       26405                  5          0   both

#### 4. Check the shape of the imported dataframe.

In [5]:
#check shape of combined orders and products
df_ords_prods.shape

(32434489, 13)

The shape matches the shape of the exported df of (32434489, 11)

In [6]:
#delete _merge column as new df
df_ords_prods_2=df_ords_prods.drop(columns=['_merge'])

In [7]:
#check output
df_ords_prods_2.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             0           0   2539329        1             1   
2             0           0   2539329        1             1   
3             0           0   2539329        1             1   
4             0           0   2539329        1             1   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                   2                  8                     NaN   
1                   2                  8                     NaN   
2                   2                  8                     NaN   
3                   2                  8                     NaN   
4                   2                  8                     NaN   

   is_first_order  product_id  add_to_cart_order  reordered  
0               1         196                  1          0  
1               1       14084                  2          0  
2               1       12427                  3          0  
3               1       26088                  4          0  
4               1       26405                  5          0

In [8]:
df_ords_prods_2.shape

(32434489, 12)

#### 5. Determine a suitable way to combine the orders_products_combined dataframe with your products data set.

In [9]:
#Import Checked products.csv
df_prods=pd.read_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'),index_col=False)

In [10]:
#Check output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [11]:
#Delete unnecessary index column 'Unnamed: 0'
df_prods=df_prods.drop(columns=['Unnamed: 0'])

In [12]:
#Check output
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [13]:
df_prods.shape

(49672, 5)

In [14]:
#test left join merge
pd.merge(df_ords_prods_2,df_prods,on=['product_id'],indicator=True)

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0                    0           0   2539329        1             1   
1                    1           1   2398795        1             2   
2                    2           2    473747        1             3   
3                    3           3   2254736        1             4   
4                    4           4    431534        1             5   
...                ...         ...       ...      ...           ...   
32404854       3360493     3360493   1320836   202557            17   
32404855       3360494     3360494     31526   202557            18   
32404856       3376038     3376038    758936   203436             1   
32404857       3376039     3376039   2745165   203436             2   
32404858       3408036     3408036   3093936   205420             1   

          orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                          2                  8                     NaN   
1                          3                  7                    15.0   
2                          3                 12                    21.0   
3                          4                  7                    29.0   
4                          4                 15                    28.0   
...                      ...                ...                     ...   
32404854                   2                 15                     1.0   
32404855                   5                 11                     3.0   
32404856                   2                  7                     NaN   
32404857                   3                  5                    15.0   
32404858                   4                 14                     NaN   

          is_first_order  product_id  add_to_cart_order  reordered  \
0                      1         196                  1          0   
1                      0         196                  1          1   
2                      0         196                  1          1   
3                      0         196                  1          1   
4                      0         196                  1          1   
...                  ...         ...                ...        ...   
32404854               0       43553                  2          1   
32404855               0       43553                  2          1   
32404856               1       42338                  4          0   
32404857               0       42338                 16          1   
32404858               1       28818                  8          0   

                           product_name  aisle_id  department_id  prices  \
0                                  Soda        77              7     9.0   
1                                  Soda        77              7     9.0   
2                                  Soda        77              7     9.0   
3                                  Soda        77              7     9.0   
4                                  Soda        77              7     9.0   
...                                 ...       ...            ...     ...   
32404854            Orange Energy Shots        64              7     3.7   
32404855            Orange Energy Shots        64              7     3.7   
32404856          Zucchini Chips, Pesto        50             19     6.9   
32404857          Zucchini Chips, Pesto        50             19     6.9   
32404858  Hot Oatmeal Multigrain Raisin       130             14    10.3   

         _merge  
0          both  
1          both  
2          both  
3          both  
4          both  
...         ...  
32404854   both  
32404855   both  
32404856   both  
32404857   both  
32404858   both  

[32404859 rows x 17 columns]

In [15]:
#Inner merge the two dataframes
df_merged_inner=pd.merge(df_ords_prods_2,df_prods,on=['product_id'],indicator=True)

In [16]:
#check inner merge output
df_merged_inner.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             1           1   2398795        1             2   
2             2           2    473747        1             3   
3             3           3   2254736        1             4   
4             4           4    431534        1             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                   2                  8                     NaN   
1                   3                  7                    15.0   
2                   3                 12                    21.0   
3                   4                  7                    29.0   
4                   4                 15                    28.0   

   is_first_order  product_id  add_to_cart_order  reordered product_name  \
0               1         196                  1          0         Soda   
1               0         196                  1          1         Soda   
2               0         196                  1          1         Soda   
3               0         196                  1          1         Soda   
4               0         196                  1          1         Soda   

   aisle_id  department_id  prices _merge  
0        77              7     9.0   both  
1        77              7     9.0   both  
2        77              7     9.0   both  
3        77              7     9.0   both  
4        77              7     9.0   both

In [17]:
#check merge shape
df_merged_inner.shape

(32404859, 17)

#### 6. Confirm the results of the merge using the merge flag.

In [18]:
#check inner merge flags
df_merged_inner[('_merge')].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### 7. Export the newly created dataframe as ords_prods_merge in a suitable format (taking into consideration the size).

In [19]:
#Export merged dataframes to pickle format
df_merged_inner.to_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_merge.pkl'))